src/version_check.py

```diff
@@ -0,0 +1,244 @@
+        relations_to_check=[x,y,z],
+        # only use if the version doesn't not need to exactly match our current version
+        version_validity_range={"x": "<a,>b"})
+
+    in update status hook [OPTIONAL]:
+    if not self.version_checker.are_related_apps_valid():
+        logger.debug(
+            "Warning relational version check failed, these relations have mismatched versions",
+            "%s",
+            self.version_checker(self.version_checker.get_invalid_versions())
+        )
+        # can set status, instruct user to change
+
+2. other areas of the charm (i.e. joined events, action events, etc) [OPTIONAL]:
+    if not self.charm.version_checker.are_related_apps_valid():
+        # do something - i.e. fail event or log message
+
+3. in upgrade handler of requirer + provider [REQUIRED]:
+    if [last unit to upgrade]:
+        self.charm.version.set_version_across_all_relations()
+
+
+"""
+import logging
+from typing import Dict, List, Optional, Tuple
+
+from ops.framework import Object
+from ops.model import Unit
+
+logger = logging.getLogger(__name__)
+
+VERSION_CONST = "version"
+DEPLOYMENT_TYPE = "deployment"
+PREFIX_DIR = "/var/lib/juju/agents/"
+LOCAL_BUILT_CHARM_PREFIX = "local"
+
+
+def get_charm_revision(unit: Unit) -> int:
+    """Returns the charm revision.
+
+    TODO: Keep this until ops framework supports: https://github.com/canonical/operator/issues/1255
+    """
+    file_path = f"{PREFIX_DIR}unit-{unit.name.replace('/','-')}/charm/.juju-charm"
+    with open(file_path) as f:
+        charm_path = f.read().rstrip()
+
+    # revision of charm in a locally built chamr is unreliable:
+    # https://chat.canonical.com/canonical/pl/ro9935ayxbyyxn9hn6opy4f4xw
+    if charm_path.split(":")[0] == LOCAL_BUILT_CHARM_PREFIX:
+        logger.debug("Charm is locally built. Cannot determine revision number.")
+        return 0
+
+    # charm_path is of the format ch:amd64/jammy/<charm-name>-<revision number>
+    revision = charm_path.split("-")[-1]
+    return int(revision)
+
+
+class CrossAppVersionChecker(Object):
+    """Verifies versions across multiple integrated applications."""
+
+    def __init__(
+        self,
+        charm,
+        version: int,
+        relations_to_check: List[str],
+        version_validity_range: Optional[Dict] = None,
+    ) -> None:
+        """Constructor for CrossAppVersionChecker object.
+
+        Args:
+            charm: charm to inherit from
+            version: (int), the current version of the desired attribute of the charm
+            relations_to_check: (List), a list of relations who should have compatible versions
+                with the current charm
+            version_validity_range: (Optional Dict), a list of ranges for valid version ranges.
+                If not provided it is assumed that relations on the provided interface must have
+                the same version.
+        """
+        self.charm = charm
+        super().__init__(self.charm, None)
```

Comment:
NIT: We can probably swap these, right? 
```suggestion
        super().__init__(charm, None)
        self.charm = charm
```